### PR TITLE
[WIP] Registration form

### PR DESCRIPTION
**Pull request by srallen**
Package: app-project

WIP start of the registration form branched off of #888. No validations or API calls yet, just the UI bits and some styling work still needs to be done. 

# Review Checklist

## General

- [ ] Are the tests passing locally and on Travis?
- [ ] Is the documentation up to date?
- [ ] Is the changelog updated?

## Apps

- [ ] Does it work in all major browsers: Firefox, Chrome, Edge, Safari?
- [ ] Does it work on mobile?
- [ ] Can you `rm -rf node_modules/ && npm install` and app works as expected?

### CODE REVIEW

File: packages/app-project/src/components/AuthModal/AuthModal.js [hunk 1]
```diff
@@ -42,6 +44,11 @@ function AuthModal (props) {
             <LoginForm closeModal={closeModal} />
           </Box>
         </Tab>
+        <Tab title='Register'>
+          <Box width='large'>
```
Comment (rogerhutchings) [inline packages/app-project/src/components/AuthModal/AuthModal.js]: I've done this for LoginModal as well, but this `Box` is unnecessary, the width prop can be set inside RegisterModal

File: packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterForm.js [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    emailListSignUp: false,
+    password: '',
+    passwordConfirm: '',
+    privacyAgreement: false,
+    realName: '',
+    username: '',
+    underageWithParent: false
+  }
+  return (
+    <Box>
```
Comment (rogerhutchings) [inline packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterForm.js]: set width here

File: packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterForm.js [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+                  checked={values.privacyAgreement}
+                  disabled={isSubmitting}
+                  id={privacyAgreementFieldId}
+                  label={<Label>{counterpart('RegisterForm.privacyAgreement')}{' '}<PrivacyPolicyLink /></Label>}
+                  onChange={handleChange}
+                />
+              </FormField>
+
+              <FormField
+                error={errors.emailListSignUp && touched.emailListSignUp && errors.emailListSignUp}
```
Comment (rogerhutchings) [inline packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterForm.js]: there won't be any errors for a non-required boolean field

File: packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterForm.js [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+                  checked={values.emailListSignUp}
+                  disabled={isSubmitting}
+                  id={emailListSignUpFieldId}
+                  label={<Label>{counterpart('RegisterForm.emailListSignUp')}</Label>}
+                  onChange={handleChange}
+                />
+              </FormField>
+
+              <FormField
+                error={errors.betaListSignUp && touched.betaListSignUp && errors.betaListSignUp}
```
Comment (rogerhutchings) [inline packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterForm.js]: there won't be any errors for a non-required boolean field

File: packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterForm.js [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+                  checked={values.betaListSignUp}
+                  disabled={isSubmitting}
+                  id={betaListSignUpFieldId}
+                  label={<Label>{counterpart('RegisterForm.betaListSignUp')}</Label>}
+                  onChange={handleChange}
+                />
+              </FormField>
+
+              <FormField
+                error={errors.underageWithParent && touched.underageWithParent && errors.underageWithParent}
```
Comment (rogerhutchings) [inline packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterForm.js]: there won't be any errors for a non-required boolean field - not sure on the legalese, but maybe this should actually be "i am over 16, or if under 16 have completed this with a parent" and we make it required

File: packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterFormContainer.js [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+  onSubmit(values, { setFieldError, setSubmitting }) {
+    const { authClient, store } = this.props
+    authClient.register(values)
+      .then(userResource => {
+        setSubmitting(false)
+        store.user.set(userResource)
+        this.props.closeModal()
+      })
+      .catch(error => {
+        setFieldError('password', error.message)
```
Comment (rogerhutchings) [inline packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterFormContainer.js]: this was set to the password field for the sign-in box as the nearest handy field help. we may need to do some parsing of the error message to cover valid email, taken username etc

File: packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterForm.js [hunk 7]
```diff
@@ -0,0 +1,235 @@
...
+      <Heading size='small' margin={{ bottom: 'xsmall', top: 'none' }}>
+        {counterpart('RegisterForm.heading')}
+      </Heading>
+      <Text>
+        {counterpart('RegisterForm.instruction')}
+      </Text>
+
+      <Formik
+        initialValues={initialValues}
+        onSubmit={onSubmit}
```
Comment (rogerhutchings) [inline packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterForm.js]: we need a validation function to check:

- passwords are identical
- username is not taken?
- email hasn't already been used?

Formik can do async validation: https://jaredpalmer.com/formik/docs/guides/validation#validate

File: packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterForm.js [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+      >
+        {({
+          errors,
+          handleChange,
+          handleBlur,
+          handleSubmit,
+          isSubmitting,
+          touched,
+          values,
+          /* and other goodies */
```
Comment (rogerhutchings) [inline packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterForm.js]: delete comment

**Comment by rogerhutchings** [comment]
Just clocked the bit about validation / api calls 🤦‍♂

**Comment by rogerhutchings** [comment]
Just saw this is branched off your review branch - I've reviewed and merged it, so I'll base it off of the original auth-modal branch

File: packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterForm.js (hunk 5 above)
Comment (srallen) [reply packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterForm.js]: This is using the current text on the registration form, which I'm assuming was reviewed by lawyers(?) at Oxford. @camallen can confirm this I think.

Comment (camallen) [reply packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterForm.js]: please copy the text verbatim from the PFE login forms, changes would need review again.

File: packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterFormContainer.js (hunk 6 above)
Comment (srallen) [reply packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterFormContainer.js]: I copied and pasted the LoginForm code as a starter :-D

File: packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterForm.js (hunk 5 above)
Comment (eatyourgreens) [reply packages/app-project/src/components/AuthModal/components/RegisterForm/RegisterForm.js]: The current registration form also adds fields (and text?) when the under-16 box is ticked. We’ll have to copy that behaviour across too, with the same wording.

**Comment by srallen** [comment]
This has a lot of merge conflicts... I'm going to try to rebase the other branch and PR adding all the async API calls and validations instead and close this.

**Pull request closed without merging**